File: WorkoutTracker/Services/EquipmentCalculator.swift
import Foundation

struct PlateConfiguration {
    let targetWeight: Double
    let barWeight: Double
    let achievedWeight: Double
    let perSide: [(plate: Double, count: Int)]
    let isLandmine: Bool

    init(targetWeight: Double, barWeight: Double, achievedWeight: Double,
         perSide: [(plate: Double, count: Int)], isLandmine: Bool = false) {
        self.targetWeight = targetWeight
        self.barWeight = barWeight
        self.achievedWeight = achievedWeight
        self.perSide = perSide
        self.isLandmine = isLandmine
    }

    var displayString: String {
        if perSide.isEmpty {
            return "Bar only (\(barWeight.cleanWeight) lbs)"
        }
        let parts = perSide.map { "\($0.count)x\($0.plate.cleanWeight)" }
        return parts.joined(separator: " + ")
    }

    var fullDisplayString: String {
        if perSide.isEmpty {
            return "Bar only (\(barWeight.cleanWeight) lbs)"
        }
        let plateStr = displayString
        let label = isLandmine ? "One side" : "Each side"
        if achievedWeight != targetWeight {
            return "\(label): \(plateStr) (\(achievedWeight.cleanWeight) lbs total)"
        }
        return "\(label): \(plateStr) (\(achievedWeight.cleanWeight) lbs)"
    }

    var weightChanged: Bool {
        achievedWeight != targetWeight
    }
}

enum EquipmentDisplay {
    case barbell(PlateConfiguration)
    case powerblock(Double)
    case dumbbell(Double)
    case band(name: String, range: String)
    case bodyweight(detail: String?)
    case assisted(weight: Double, detail: String?)
    case kettlebell(Double)

    var text: String {
        switch self {
        case .barbell(let config):
            return config.fullDisplayString
        case .powerblock(let weight):
            return "PowerBlock: \(weight.cleanWeight) lbs"
        case .dumbbell(let weight):
            return "Dumbbell: \(weight.cleanWeight) lbs"
        case .band(let name, let range):
            return "\(name) Band (\(range))"
        case .bodyweight(let detail):
            if let detail, !detail.isEmpty {
                return detail
            }
            return "Bodyweight"
        case .assisted(let weight, let detail):
            if weight > 0 {
                if let detail {
                    return "\(detail) (~\(weight.cleanWeight) lb assistance)"
                }
                return "\(weight.cleanWeight) lb assistance"
            }
            return detail ?? "Assisted"
        case .kettlebell(let weight):
            return "Kettlebell: \(weight.cleanWeight) lbs"
        }
    }

    var shortText: String {
        switch self {
        case .barbell(let config):
            return "\(config.achievedWeight.cleanWeight) lbs"
        case .powerblock(let w), .dumbbell(let w), .kettlebell(let w):
            return "\(w.cleanWeight) lbs"
        case .band(let name, _):
            return "\(name) Band"
        case .bodyweight:
            return "BW"
        case .assisted(let weight, _):
            return weight > 0 ? "\(weight.cleanWeight) lb assist" : "Assisted"
        }
    }
}

struct EquipmentCalculator {
    // Available plates per side (weight, max count per side)
    var availablePlates: [(weight: Double, maxPerSide: Int)] = [
        (45, 1),
        (35, 1),
        (25, 2),
        (10, 2),
        (5, 1),
        (2.5, 1),
        (1, 1),
        (0.75, 1),
        (0.5, 1)
    ]

    static let bandInfo: [String: (name: String, range: String)] = [
        "Orange": (name: "Orange", range: "2-12 lbs"),
        "Purple": (name: "Purple", range: "5-35 lbs"),
        "Red": (name: "Red", range: "10-50 lbs"),
        "Blue": (name: "Blue", range: "20-80 lbs"),
    ]

    // MARK: - Barbell Plate Calculator

    func calculateBarbell(targetWeight: Double, barWeight: Double) -> PlateConfiguration {
        let perSideNeeded = (targetWeight - barWeight) / 2.0

        if perSideNeeded <= 0 {
            return PlateConfiguration(
                targetWeight: targetWeight,
                barWeight: barWeight,
                achievedWeight: barWeight,
                perSide: []
            )
        }

        // Try exact match first
        if let exact = findPlates(target: perSideNeeded) {
            let achieved = barWeight + exact.totalPerSide * 2
            return PlateConfiguration(
                targetWeight: targetWeight,
                barWeight: barWeight,
                achievedWeight: achieved,
                perSide: exact.plates
            )
        }

        // Round up: try incrementally higher weights until we find one
        var attempt = perSideNeeded
        let increment = 0.25 // smallest meaningful increment
        while attempt < perSideNeeded + 50 {
            attempt += increment
            if let found = findPlates(target: attempt) {
                let achieved = barWeight + found.totalPerSide * 2
                return PlateConfiguration(
                    targetWeight: targetWeight,
                    barWeight: barWeight,
                    achievedWeight: achieved,
                    perSide: found.plates
                )
            }
        }

        // Fallback: just the bar
        return PlateConfiguration(
            targetWeight: targetWeight,
            barWeight: barWeight,
            achievedWeight: barWeight,
            perSide: []
        )
    }

    private func findPlates(target: Double) -> (plates: [(plate: Double, count: Int)], totalPerSide: Double)? {
        var remaining = target
        var result: [(plate: Double, count: Int)] = []
        let epsilon = 0.001

        for plate in availablePlates {
            if remaining < plate.weight - epsilon { continue }

            let maxCount = min(plate.maxPerSide, Int(remaining / plate.weight))
            if maxCount > 0 {
                result.append((plate.weight, maxCount))
                remaining -= Double(maxCount) * plate.weight
            }
        }

        if abs(remaining) < epsilon {
            let total = result.reduce(0.0) { $0 + Double($1.count) * $1.plate }
            return (result, total)
        }

        return nil
    }

    func calculateLandmine(targetWeight: Double, barWeight: Double) -> PlateConfiguration {
        let oneSideNeeded = targetWeight - barWeight

        if oneSideNeeded <= 0 {
            return PlateConfiguration(targetWeight: targetWeight, barWeight: barWeight,
                                      achievedWeight: barWeight, perSide: [], isLandmine: true)
        }

        if let exact = findPlates(target: oneSideNeeded) {
            let achieved = barWeight + exact.totalPerSide
            return PlateConfiguration(targetWeight: targetWeight, barWeight: barWeight,
                                      achievedWeight: achieved, perSide: exact.plates, isLandmine: true)
        }

        var attempt = oneSideNeeded
        let increment = 0.25
        while attempt < oneSideNeeded + 50 {
            attempt += increment
            if let found = findPlates(target: attempt) {
                let achieved = barWeight + found.totalPerSide
                return PlateConfiguration(targetWeight: targetWeight, barWeight: barWeight,
                                          achievedWeight: achieved, perSide: found.plates, isLandmine: true)
            }
        }

        return PlateConfiguration(targetWeight: targetWeight, barWeight: barWeight,
                                  achievedWeight: barWeight, perSide: [], isLandmine: true)
    }

    private func isLandmine(_ exercise: Exercise) -> Bool {
        exercise.name.lowercased().contains("landmine")
    }

    // MARK: - PowerBlock Calculator

    func nearestPowerBlock(target: Double) -> Double {
        let clamped = max(5, min(50, target))
        return (clamped / 2.5).rounded() * 2.5
    }

    // MARK: - Equipment Display

    func display(for exercise: Exercise, weight: Double) -> EquipmentDisplay {
        switch exercise.equipmentType {
        case .barbell_45:
            let landmine45 = isLandmine(exercise)
            if weight <= 0 || weight <= 45 {
                return .barbell(PlateConfiguration(targetWeight: 45, barWeight: 45, achievedWeight: 45, perSide: [], isLandmine: landmine45))
            }
            return .barbell(landmine45 ? calculateLandmine(targetWeight: weight, barWeight: 45)
                                       : calculateBarbell(targetWeight: weight, barWeight: 45))

        case .barbell_35:
            let landmine35 = isLandmine(exercise)
            if weight <= 0 || weight <= 35 {
                return .barbell(PlateConfiguration(targetWeight: 35, barWeight: 35, achievedWeight: 35, perSide: [], isLandmine: landmine35))
            }
            return .barbell(landmine35 ? calculateLandmine(targetWeight: weight, barWeight: 35)
                                       : calculateBarbell(targetWeight: weight, barWeight: 35))

        case .barbell_ez:
            if weight <= 0 || weight <= 15 {
                return .barbell(PlateConfiguration(targetWeight: 15, barWeight: 15, achievedWeight: 15, perSide: []))
            }
            return .barbell(calculateBarbell(targetWeight: weight, barWeight: 15))

        case .powerblock:
            // Check if this is a regular dumbbell (below PowerBlock range)
            if let detail = exercise.equipmentDetail,
               let dbWeight = parseWeight(from: detail), dbWeight < 5 {
                return .dumbbell(dbWeight)
            }
            if weight > 0 && weight < 5 {
                return .dumbbell(weight)
            }
            return .powerblock(weight > 0 ? nearestPowerBlock(target: weight) : 0)

        case .band:
            let bandName = exercise.equipmentDetail ?? "Unknown"
            let info = Self.bandInfo[bandName]
            return .band(name: info?.name ?? bandName, range: info?.range ?? "")

        case .bodyweight:
            return .bodyweight(detail: exercise.equipmentDetail)

        case .assisted_pullup:
            return .assisted(weight: weight, detail: exercise.equipmentDetail)

        case .kettlebell:
            return .kettlebell(weight)
        }
    }

    /// Difference display between two exercises (for weight change alerts)
    func weightChangeDisplay(from previousWeight: Double, to currentWeight: Double, exercise: Exercise) -> String? {
        guard let barWeight = exercise.equipmentType.barWeight else { return nil }
        let landmine = isLandmine(exercise)
        let divisor = landmine ? 1.0 : 2.0
        let prevSide = (previousWeight - barWeight) / divisor
        let currSide = (currentWeight - barWeight) / divisor
        let diff = currSide - prevSide

        if abs(diff) < 0.001 { return nil }

        let sideLabel = landmine ? "one side" : "each side"
        if diff > 0 {
            return "ADD \(diff.cleanWeight) lbs to \(sideLabel)"
        } else {
            return "REMOVE \(abs(diff).cleanWeight) lbs from \(sideLabel)"
        }
    }

    private func parseWeight(from detail: String) -> Double? {
        let cleaned = detail.replacingOccurrences(of: "lb", with: "")
            .replacingOccurrences(of: "s", with: "")
            .trimmingCharacters(in: .whitespaces)
        return Double(cleaned)
    }
}

// MARK: - Double Extension

extension Double {
    var cleanWeight: String {
        if self == self.rounded() {
            return String(format: "%.0f", self)
        }
        // Remove trailing zeros but keep necessary decimals
        let formatted = String(format: "%.2f", self)
        var result = formatted
        while result.hasSuffix("0") { result.removeLast() }
        if result.hasSuffix(".") { result.removeLast() }
        return result
    }
}

File: web/src/lib/equipment-calculator.ts
import type { Exercise, PlateConfiguration, EquipmentDisplay, PowerBlockInstructions, UserEquipmentConfig } from "./types";
import { barWeight, cleanWeight, ALL_BAND_COLORS, ALL_LOOP_BAND_SIZES, ALL_COC_LEVELS } from "./types";

export const FIXED_DUMBBELLS: number[] = [2];
export const KETTLEBELL_WEIGHTS: number[] = [15, 25, 45];

// totalOwned = total physical plates owned across both sides.
// For symmetric barbell: max per side = floor(totalOwned / 2).
// For landmine (one-sided): all plates can go on the single loaded end.
export const DEFAULT_PLATES: { weight: number; totalOwned: number }[] = [
  { weight: 45, totalOwned: 2 },
  { weight: 35, totalOwned: 2 },
  { weight: 25, totalOwned: 4 },
  { weight: 10, totalOwned: 4 },
  { weight: 5, totalOwned: 2 },
  { weight: 2.5, totalOwned: 2 },
  { weight: 1, totalOwned: 2 },
  { weight: 0.75, totalOwned: 2 },
  { weight: 0.5, totalOwned: 2 },
];

export const DEFAULT_EQUIPMENT_CONFIG: UserEquipmentConfig = {
  barbells: { has45lb: true, has35lb: true, hasEZBar: true },
  plates: DEFAULT_PLATES.map((p) => ({ ...p })),
  powerBlock: { owned: true, minLbs: 5, maxLbs: 50 },
  fixedDumbbells: [...FIXED_DUMBBELLS],
  kettlebells: [...KETTLEBELL_WEIGHTS],
  bands: [...ALL_BAND_COLORS],
  loopBands: [...ALL_LOOP_BAND_SIZES],
  assistedPullupBands: 2,
  grippers: [...ALL_COC_LEVELS],
};

const BAND_INFO: Record<string, { name: string; range: string }> = {
  Orange: { name: "Orange", range: "2-12 lbs" },
  Purple: { name: "Purple", range: "5-35 lbs" },
  Red: { name: "Red", range: "10-50 lbs" },
  Blue: { name: "Blue", range: "20-80 lbs" },
  Green: { name: "Green", range: "50-120 lbs" },
  Black: { name: "Black", range: "60-150 lbs" },
};

// ── Plate Calculator ──

function getEffectivePlates(
  config?: UserEquipmentConfig
): { weight: number; totalOwned: number }[] {
  const plates = config?.plates ?? DEFAULT_PLATES;
  // Defensive: greedy algorithm assumes largest → smallest order.
  return [...plates].sort((a, b) => b.weight - a.weight);
}

// loadingSides: 2 for symmetric barbell (plates on both ends), 1 for landmine (one end only).
// Per-side limit = floor(totalOwned / loadingSides).
function findPlates(
  target: number,
  plates: { weight: number; totalOwned: number }[] = DEFAULT_PLATES,
  loadingSides: 1 | 2 = 2
): { plates: { plate: number; count: number }[]; totalPerSide: number } | null {
  let remaining = target;
  const result: { plate: number; count: number }[] = [];
  const epsilon = 0.001;

  for (const plate of plates) {
    if (remaining < plate.weight - epsilon) continue;
    const maxPerSide = Math.floor(plate.totalOwned / loadingSides);
    const maxCount = Math.min(maxPerSide, Math.floor(remaining / plate.weight));
    if (maxCount > 0) {
      result.push({ plate: plate.weight, count: maxCount });
      remaining -= maxCount * plate.weight;
    }
  }

  if (Math.abs(remaining) < epsilon) {
    const total = result.reduce((sum, p) => sum + p.count * p.plate, 0);
    return { plates: result, totalPerSide: total };
  }
  return null;
}

function isLandmineExercise(exercise: Exercise): boolean {
  return exercise.name.toLowerCase().includes("landmine");
}

export function calculateBarbell(targetWeight: number, bWeight: number, config?: UserEquipmentConfig): PlateConfiguration {
  const plates = getEffectivePlates(config);
  const perSideNeeded = (targetWeight - bWeight) / 2;

  if (perSideNeeded <= 0) {
    return { targetWeight, barWeight: bWeight, achievedWeight: bWeight, perSide: [] };
  }

  // Try exact match
  const exact = findPlates(perSideNeeded, plates, 2);
  if (exact) {
    return {
      targetWeight,
      barWeight: bWeight,
      achievedWeight: bWeight + exact.totalPerSide * 2,
      perSide: exact.plates,
    };
  }

  // Round down: find nearest achievable weight below target
  const step = 0.25;
  let tryPerSide = Math.floor((perSideNeeded - 0.001) / step) * step;
  while (tryPerSide > 0) {
    const normalized = Math.round(tryPerSide * 1000) / 1000;
    const found = findPlates(normalized, plates, 2);
    if (found) {
      return {
        targetWeight,
        barWeight: bWeight,
        achievedWeight: bWeight + found.totalPerSide * 2,
        perSide: found.plates,
      };
    }
    tryPerSide -= step;
  }

  return { targetWeight, barWeight: bWeight, achievedWeight: bWeight, perSide: [] };
}

export function calculateLandmine(targetWeight: number, bWeight: number, config?: UserEquipmentConfig): PlateConfiguration {
  const plates = getEffectivePlates(config);
  const oneSideNeeded = targetWeight - bWeight;

  if (oneSideNeeded <= 0) {
    return { targetWeight, barWeight: bWeight, achievedWeight: bWeight, perSide: [], isLandmine: true };
  }

  // Try exact match (loadingSides=1: all owned plates available on the single loaded end)
  const exact = findPlates(oneSideNeeded, plates, 1);
  if (exact) {
    return {
      targetWeight,
      barWeight: bWeight,
      achievedWeight: bWeight + exact.totalPerSide,
      perSide: exact.plates,
      isLandmine: true,
    };
  }

  // Round down: find nearest achievable weight below target
  const step = 0.25;
  let tryOneSide = Math.floor((oneSideNeeded - 0.001) / step) * step;
  while (tryOneSide > 0) {
    const normalized = Math.round(tryOneSide * 1000) / 1000;
    const found = findPlates(normalized, plates, 1);
    if (found) {
      return {
        targetWeight,
        barWeight: bWeight,
        achievedWeight: bWeight + found.totalPerSide,
        perSide: found.plates,
        isLandmine: true,
      };
    }
    tryOneSide -= step;
  }

  return { targetWeight, barWeight: bWeight, achievedWeight: bWeight, perSide: [], isLandmine: true };
}

// ── PowerBlock ──

export function nearestPowerBlock(target: number, config?: UserEquipmentConfig): number {
  const min = config?.powerBlock?.minLbs ?? 5;
  const max = config?.powerBlock?.maxLbs ?? 50;
  const step = 2.5;
  const snapped = Math.round(target / step) * step;
  return Math.max(min, Math.min(max, snapped));
}

// PowerBlock Elite EXP Stage 1 selector + rod configuration per weight
const POWERBLOCK_TABLE: Record<number, { selector: number; rods: "none" | "one" | "both" }> = {
  5:    { selector: 5,  rods: "none" },
  7.5:  { selector: 5,  rods: "one" },
  10:   { selector: 5,  rods: "both" },
  12.5: { selector: 10, rods: "none" },
  15:   { selector: 10, rods: "one" },
  17.5: { selector: 10, rods: "both" },
  20:   { selector: 15, rods: "none" },
  22.5: { selector: 15, rods: "one" },
  25:   { selector: 25, rods: "none" },
  27.5: { selector: 25, rods: "one" },
  30:   { selector: 25, rods: "both" },
  32.5: { selector: 30, rods: "none" },
  35:   { selector: 30, rods: "one" },
  37.5: { selector: 30, rods: "both" },
  40:   { selector: 40, rods: "none" },
  42.5: { selector: 40, rods: "one" },
  45:   { selector: 40, rods: "both" },
  47.5: { selector: 45, rods: "one" },
  50:   { selector: 50, rods: "none" },
};

const ROD_LABEL: Record<"none" | "one" | "both", string> = {
  none: "no rods",
  one: "add 1 rod",
  both: "add both rods",
};

export function getPowerBlockInstructions(weight: number): PowerBlockInstructions {
  const entry = POWERBLOCK_TABLE[weight];
  if (!entry) {
    return { selector: 5, rods: "none", label: "—" };
  }
  const { selector, rods } = entry;
  return { selector, rods, label: `Selector to ${selector} · ${ROD_LABEL[rods]}` };
}

// ── Display helpers ──

export function plateDisplayString(config: PlateConfiguration): string {
  if (config.perSide.length === 0) {
    return `Bar only (${cleanWeight(config.barWeight)} lbs)`;
  }
  return config.perSide.map((p) => `${p.count}x${cleanWeight(p.plate)}`).join(" + ");
}

function plateBracketString(config: PlateConfiguration): string {
  const items = config.perSide.map((p) => `${p.count}x${cleanWeight(p.plate)}`).join(", ");
  return `[${items}]`;
}

export function plateFullDisplayString(config: PlateConfiguration): string {
  if (config.perSide.length === 0) {
    return `Bar only (${cleanWeight(config.barWeight)} lbs)`;
  }
  const brackets = plateBracketString(config);
  const label = config.isLandmine ? "one end" : "each side";
  return `${cleanWeight(config.barWeight)}lb bar + ${brackets} ${label} (${cleanWeight(config.achievedWeight)} lbs)`;
}

function parseWeightFromDetail(detail: string): number | null {
  const cleaned = detail.replace(/lbs?/gi, "").replace(/s$/i, "").trim();
  const num = parseFloat(cleaned);
  return isNaN(num) ? null : num;
}

export function getEquipmentDisplay(exercise: Exercise, weight: number, config?: UserEquipmentConfig): EquipmentDisplay {
  switch (exercise.equipmentType) {
    case "barbell_45": {
      const landmine = isLandmineExercise(exercise);
      if (weight <= 45) return { type: "barbell", config: { targetWeight: 45, barWeight: 45, achievedWeight: 45, perSide: [], isLandmine: landmine } };
      return { type: "barbell", config: landmine ? calculateLandmine(weight, 45, config) : calculateBarbell(weight, 45, config) };
    }

    case "barbell_35": {
      const landmine = isLandmineExercise(exercise);
      if (weight <= 35) return { type: "barbell", config: { targetWeight: 35, barWeight: 35, achievedWeight: 35, perSide: [], isLandmine: landmine } };
      return { type: "barbell", config: landmine ? calculateLandmine(weight, 35, config) : calculateBarbell(weight, 35, config) };
    }

    case "barbell_ez":
      if (weight <= 15) return { type: "barbell", config: { targetWeight: 15, barWeight: 15, achievedWeight: 15, perSide: [] } };
      return { type: "barbell", config: calculateBarbell(weight, 15, config) };

    case "powerblock": {
      if (exercise.equipmentDetail) {
        const dbWeight = parseWeightFromDetail(exercise.equipmentDetail);
        if (dbWeight !== null && dbWeight < 5) return { type: "dumbbell", weight: dbWeight };
      }
      // If PowerBlock not owned, treat as plain dumbbell
      if (config?.powerBlock?.owned === false) return { type: "dumbbell", weight };
      if (weight > 0 && weight < (config?.powerBlock?.minLbs ?? 5)) return { type: "dumbbell", weight };
      const resolvedWeight = weight > 0 ? nearestPowerBlock(weight, config) : 0;
      const instructions = resolvedWeight > 0
        ? getPowerBlockInstructions(resolvedWeight)
        : { selector: 5, rods: "none" as const, label: "—" };
      return { type: "powerblock", weight: resolvedWeight, instructions };
    }

    case "dumbbell":
      return { type: "dumbbell", weight };

    case "band": {
      const bandName = exercise.equipmentDetail || "Unknown";
      const info = BAND_INFO[bandName];
      return { type: "band", name: info?.name || bandName, range: info?.range || "" };
    }

    case "bodyweight":
      return { type: "bodyweight", detail: exercise.equipmentDetail };

    case "assisted_pullup":
      return { type: "assisted", weight, detail: exercise.equipmentDetail };

    case "kettlebell":
      return { type: "kettlebell", weight };

    case "gripper":
      return { type: "gripper", weight };
  }
}

export function equipmentDisplayText(display: EquipmentDisplay): string {
  switch (display.type) {
    case "barbell": return plateFullDisplayString(display.config);
    case "powerblock": return `PowerBlock: ${cleanWeight(display.weight)} lbs`;
    case "dumbbell": return `Dumbbell: ${cleanWeight(display.weight)} lbs`;
    case "band": return `${display.name} Band (${display.range})`;
    case "bodyweight": return display.detail || "Bodyweight";
    case "assisted": {
      if (display.weight > 0) {
        const bandLabel = display.weight === 1 ? "band" : "bands";
        return `${display.weight} ${bandLabel} (~${display.weight * 80} lbs assistance)`;
      }
      return display.detail?.replace(/_/g, " ") || "Assisted";
    }
    case "kettlebell": return `Kettlebell: ${cleanWeight(display.weight)} lbs`;
    case "gripper": return display.weight > 0 ? `Gripper: ${cleanWeight(display.weight)} lbs` : "Gripper";
  }
}

export function equipmentShortText(display: EquipmentDisplay): string {
  switch (display.type) {
    case "barbell": return `${cleanWeight(display.config.achievedWeight)} lbs`;
    case "powerblock":
    case "dumbbell":
    case "kettlebell":
    case "gripper": return `${cleanWeight(display.weight)} lbs`;
    case "band": return `${display.name} Band`;
    case "bodyweight": return "BW";
    case "assisted": {
      if (display.weight > 0) {
        const bandLabel = display.weight === 1 ? "band" : "bands";
        return `${display.weight} ${bandLabel}`;
      }
      return "Assisted";
    }
  }
}
